File: src/communications/email-service/app/store.py
from __future__ import annotations

import asyncio
import logging
from collections import OrderedDict
from typing import Dict, List, Optional

from .models import EmailRecord

logger = logging.getLogger(__name__)


class EmailStore:
    """Thread-safe in-memory store for EmailRecord objects.

    Keeps at most *max_size* entries; when the limit is reached the oldest
    entry (by insertion order) is evicted before the new one is saved.
    """

    def __init__(self, max_size: int = 10_000) -> None:
        self._max_size = max_size
        self._records: OrderedDict[str, EmailRecord] = OrderedDict()
        self._lock = asyncio.Lock()
        self._total_sent: int = 0
        self._total_delivered: int = 0
        self._total_failed: int = 0

    # ------------------------------------------------------------------
    # Write
    # ------------------------------------------------------------------

    async def save(self, record: EmailRecord) -> None:
        """Persist a record, evicting the oldest entry when at capacity."""
        async with self._lock:
            # Evict oldest entries until we are under the cap
            while len(self._records) >= self._max_size:
                evicted_key, _ = self._records.popitem(last=False)
                logger.debug("Evicted oldest record: %s", evicted_key)

            self._records[record.messageId] = record
            self._total_sent += 1
            if record.status == "delivered":
                self._total_delivered += 1
            else:
                self._total_failed += 1

    # ------------------------------------------------------------------
    # Read
    # ------------------------------------------------------------------

    async def get(self, message_id: str) -> Optional[EmailRecord]:
        """Return the record for *message_id*, or ``None`` if not found."""
        async with self._lock:
            return self._records.get(message_id)

    async def list(self, limit: int = 50) -> List[EmailRecord]:
        """Return up to *limit* most-recently-saved records (newest first)."""
        async with self._lock:
            items = list(self._records.values())
        # Reverse to return newest first without mutating the underlying dict
        return list(reversed(items))[:limit]

    async def get_stats(self) -> Dict[str, int]:
        """Return aggregate delivery statistics."""
        async with self._lock:
            return {
                "sent": self._total_sent,
                "delivered": self._total_delivered,
                "failed": self._total_failed,
            }


# Module-level singleton used by the API and consumer
email_store = EmailStore()

File: src/communications/push-notification-service/internal/store/store.go
// Package store provides an in-memory store for PushRecord objects.
package store

import (
	"sync"

	"github.com/shopos/push-notification-service/internal/domain"
)

// Store is a bounded, thread-safe in-memory collection of PushRecords.
// When the store reaches maxSize it evicts the oldest entry before inserting a
// new one, preserving insertion order via a slice-based key list.
type Store struct {
	mu      sync.RWMutex
	records map[string]domain.PushRecord
	order   []string // insertion-order tracking for eviction
	maxSize int

	// aggregate counters
	sent      int
	delivered int
	failed    int
}

// New creates a Store with the given maximum capacity.
func New(maxSize int) *Store {
	return &Store{
		records: make(map[string]domain.PushRecord, maxSize),
		order:   make([]string, 0, maxSize),
		maxSize: maxSize,
	}
}

// Save persists a PushRecord, evicting the oldest entry when at capacity.
func (s *Store) Save(r domain.PushRecord) {
	s.mu.Lock()
	defer s.mu.Unlock()

	// Evict oldest until we have room
	for len(s.records) >= s.maxSize {
		oldest := s.order[0]
		s.order = s.order[1:]
		delete(s.records, oldest)
	}

	s.records[r.MessageID] = r
	s.order = append(s.order, r.MessageID)

	s.sent++
	if r.Status == "delivered" {
		s.delivered++
	} else {
		s.failed++
	}
}

// Get returns the PushRecord for messageID and a boolean indicating whether it
// was found.
func (s *Store) Get(messageID string) (domain.PushRecord, bool) {
	s.mu.RLock()
	defer s.mu.RUnlock()
	r, ok := s.records[messageID]
	return r, ok
}

// List returns up to limit records in reverse-insertion order (newest first).
func (s *Store) List(limit int) []domain.PushRecord {
	s.mu.RLock()
	defer s.mu.RUnlock()

	n := len(s.order)
	if limit > n {
		limit = n
	}

	out := make([]domain.PushRecord, 0, limit)
	for i := n - 1; i >= 0 && len(out) < limit; i-- {
		if r, ok := s.records[s.order[i]]; ok {
			out = append(out, r)
		}
	}
	return out
}

// Stats returns a snapshot of aggregate delivery counters.
func (s *Store) Stats() domain.PushStats {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return domain.PushStats{
		Sent:      s.sent,
		Delivered: s.delivered,
		Failed:    s.failed,
	}
}
